libs/wire-subsystems/src/Wire/InvitationCodeStore/Cassandra.hs

```diff
@@ -0,0 +1,192 @@
+    cql =
+      [sql|
+      SELECT  team, id, code FROM team_invitation_info WHERE code = ?
+      |]
+
+lookupInvitationCodesByEmailImpl :: EmailAddress -> Client [StoredInvitationInfo]
+lookupInvitationCodesByEmailImpl email = map asRecord <$> retry x1 (query cql (params LocalQuorum (Identity email)))
+  where
+    cql :: PrepQuery R (Identity EmailAddress) (TeamId, InvitationId, InvitationCode)
+    cql =
+      [sql|
+      SELECT team, invitation, code FROM team_invitation_email WHERE email = ?
+      |]
+
+lookupInvitationImpl :: TeamId -> InvitationId -> Client (Maybe StoredInvitation)
+lookupInvitationImpl tid iid =
+  fmap asRecord
+    <$> retry x1 (query1 cql (params LocalQuorum (tid, iid)))
+  where
+    cql :: PrepQuery R (TeamId, InvitationId) (TupleType StoredInvitation)
+    cql =
+      [sql|
+      SELECT team, role, id, created_at, created_by, email, name, code FROM team_invitation WHERE team = ? AND id = ?
+      |]
+
+deleteInvitationImpl :: TeamId -> InvitationId -> Client ()
+deleteInvitationImpl teamId invId = do
+  codeEmail <- lookupInvitationCodeEmail
+  case codeEmail of
+    Just (invCode, invEmail) -> retry x5 . batch $ do
+      setType BatchLogged
+      setConsistency LocalQuorum
+      addPrepQuery cqlInvitation (teamId, invId)
+      addPrepQuery cqlInvitationInfo (Identity invCode)
+      addPrepQuery cqlInvitationEmail (invEmail, teamId)
+    Nothing ->
+      retry x5 $ write cqlInvitation (params LocalQuorum (teamId, invId))
+  where
+    lookupInvitationCodeEmail :: Client (Maybe (InvitationCode, EmailAddress))
+    lookupInvitationCodeEmail = retry x1 (query1 cqlInvitationCodeEmail (params LocalQuorum (teamId, invId)))
+
+    cqlInvitation :: PrepQuery W (TeamId, InvitationId) ()
+    cqlInvitation =
+      [sql|
+        DELETE FROM team_invitation where team = ? AND id = ?
+      |]
+
+    cqlInvitationInfo :: PrepQuery W (Identity InvitationCode) ()
+    cqlInvitationInfo =
+      [sql|
+        DELETE FROM team_invitation_info WHERE code = ?
+      |]
+
+    cqlInvitationEmail :: PrepQuery W (EmailAddress, TeamId) ()
+    cqlInvitationEmail =
+      [sql|
+        DELETE FROM team_invitation_email WHERE email = ? AND team = ?
+      |]
+
+    cqlInvitationCodeEmail :: PrepQuery R (TeamId, InvitationId) (InvitationCode, EmailAddress)
+    cqlInvitationCodeEmail =
+      [sql|
+        SELECT code, email FROM team_invitation WHERE team = ? AND id = ?
+      |]
+
+deleteInvitationsImpl :: TeamId -> Client ()
+deleteInvitationsImpl teamId =
+  runConduit $
+    paginateC cqlSelect (paramsP LocalQuorum (Identity teamId) 100) x1
+      .| Conduit.mapM_ (pooledMapConcurrentlyN_ 16 (deleteInvitationImpl teamId . runIdentity))
+  where
+    cqlSelect :: PrepQuery R (Identity TeamId) (Identity InvitationId)
+    cqlSelect = "SELECT id FROM team_invitation WHERE team = ? ORDER BY id ASC"
+
+mkInvitationCode :: IO InvitationCode
+mkInvitationCode = InvitationCode . encodeBase64Url <$> randBytes 24
```

Comment:
https://hoogle.zinfra.io/file/nix/store/ll4lg104j02cpv5li2zcx603rbdkd8gd-polysemy-wire-zoo-0.1.0-doc/share/doc/polysemy-wire-zoo-0.1.0/html/Wire-Sem-Random.html

This will also resolve our issues with `unsafePerformIO`. `unsafePerformIO` is dangerous there because, depending on optimizations, profiling, ghc version and other factors, GHC could decide to e.g. float out the `let` binding and hence not create a new invitation code.

Reply:
Seems like a neat FUTUREWORK for when we do invitation subsystems.

Reply:
`unsafePerformIO` has valid usecases but willingly breaking referential transparency is not one of them. If you only want one invitation code, put it in a separate function, put a `NOINLINE` pragma but if you want new ones every time, then it has to be in `IO`. I don't get why this is problematic to do, it's a small change and we already have all the glue ready.

Reply:
We only use `unsafePerformIO` in tests; in fact, the code where it is used is weed. Also there is a comment in the unsafe line that we should check that this actually works.

Good enough for now?

Reply:
The problem is that there is no "actually works" here since GHC might just decide differently depending on how it's configured or just randomly in some future ghc version. 

If it's weed, then why include it in the PR? I was always asked to just stash stuff that is not used anyway.